### PR TITLE
For mac and windows skip passing client storage to avoid pickling error in multiprocessing

### PR DESCRIPTION
**Pull request by akansha1812**
Fixes  https://github.com/GoogleCloudPlatform/dataflux-pytorch/issues/58

The pickling error was due to passing storage client to pytorch dataloader. When num_workers>0, pytorch uses multi-processing. Multi-processing fails as it cannot serialize/pickle storage client and lambda function. 

- [x] Tests pass
- [x] Appropriate changes to documentation are included in the PR

### CODE REVIEW

File: dataflux_pytorch/dataflux_iterable_dataset.py [hunk 1]
```diff
@@ -111,53 +114,66 @@ def __init__(
                 during initialization.
         """
         super().__init__()
-        self.storage_client = storage_client
-        if not storage_client:
-            self.storage_client = storage.Client(project=project_name, )
-        dataflux_core.user_agent.add_dataflux_user_agent(self.storage_client)
+        self.storage_client = None
+        if OS == LINUX:
```
Comment (MattIrv) [inline dataflux_pytorch/dataflux_iterable_dataset.py:118]: It looks like this means we're silently ignoring the passed-in storage client when the user is on other platforms? If so, should we either raise an exception or log a warning when it's set on other platforms?

In fact, it seems like in some cases it may work for the user and for you on Mac, but it's not entirely clear when, is that right? If so we may want to log a warning saying that setting the storage client is not supported on non-Linux platforms but try to do it anyway

File: dataflux_pytorch/dataflux_iterable_dataset.py [hunk 2]
```diff
@@ -111,53 +114,66 @@ def __init__(
...
+                storage_client
+                if storage_client is not None
+                else storage.Client(project=project_name)
+            )
+            dataflux_core.user_agent.add_dataflux_user_agent(self.storage_client)
         self.project_name = project_name
         self.bucket_name = bucket_name
-        self.data_format_fn = data_format_fn
+        self.data_format_fn = (
+            data_format_fn if data_format_fn is not None else self.data_format_default
```
Comment (MattIrv) [inline dataflux_pytorch/dataflux_iterable_dataset.py:128]: If the user passes a custom data_format_fn which is a lambda, will it fail to pickle it? If so, should we add a doc comment or something that on non-Linux platforms they need to pass a function definition rather than a lambda?

Comment (MattIrv) [inline dataflux_pytorch/dataflux_iterable_dataset.py:128]: I think it may be possible to define data_format_default inline even using "def" syntax rather than maxing it a method on the class

File: dataflux_pytorch/dataflux_iterable_dataset.py [hunk 3]
```diff
@@ -111,53 +114,66 @@ def __init__(
...
-                        dataflux_core.download.dataflux_download_lazy(
-                            project_name=self.project_name,
-                            bucket_name=self.bucket_name,
-                            objects=self.objects,
-                            storage_client=self.storage_client,
-                            dataflux_download_optimization_params=self.
-                            dataflux_download_optimization_params,
-                            retry_config=self.config.download_retry_config,
-                        ))
+            yield from (
```
Comment (MattIrv) [inline dataflux_pytorch/dataflux_iterable_dataset.py:143]: Here and elsewhere in this PR, it looks like your formatter might have reformatted some things with different settings than we're using for the project. Can you make sure that the recommended extensions are installed (Extensions > Recommended in VS Code) and then re-run the formatter on these files?

File: dataflux_pytorch/dataflux_mapstyle_dataset.py [hunk 4]
```diff
@@ -115,18 +118,25 @@ def __init__(
             None.
         """
         super().__init__()
-        self.storage_client = storage_client
-        if not storage_client:
-            self.storage_client = storage.Client(project=project_name, )
-        dataflux_core.user_agent.add_dataflux_user_agent(self.storage_client)
+        self.storage_client = None
+        if OS == LINUX:
```
Comment (MattIrv) [inline dataflux_pytorch/dataflux_mapstyle_dataset.py:122]: Same comments on this file as above

File: dataflux_pytorch/dataflux_mapstyle_dataset.py [hunk 5]
```diff
@@ -140,23 +150,26 @@ def __getitem__(self, idx):
...
                 retry_config=self.config.download_retry_config,
             )
         ]
 
+    def data_format_default(self, data):
+        return data
+
     def _list_GCS_blobs_with_retry(self):
         """Retries Dataflux Listing upon exceptions, up to the retries defined in self.config."""
         error = None
```
Comment (MattIrv) [inline dataflux_pytorch/dataflux_mapstyle_dataset.py:175]: There are no tests in this PR, is it possible to add tests that catch these failures, for example by creating dataset objects and trying to pickle them?

File: dataflux_pytorch/dataflux_iterable_dataset.py (hunk 1 above)
Comment (jdnurme) [reply dataflux_pytorch/dataflux_iterable_dataset.py:118]: Definitely log a warning at very least to let them know it's being ignored. I would even say throwing an error is probably appropriate so they just remove the setting.

Comment (jdnurme) [reply dataflux_pytorch/dataflux_iterable_dataset.py:118]: Mac is unstable with fork (see https://docs.python.org/3/library/multiprocessing.html#contexts-and-start-methods) so safest option is to move forward with linux only.

File: dataflux_pytorch/dataflux_iterable_dataset.py (hunk 2 above)
Comment (jdnurme) [reply dataflux_pytorch/dataflux_iterable_dataset.py:128]: Lambda is definitely preferable for readability unless you ran into specific errors with passing a lambda.

File: dataflux_pytorch/dataflux_mapstyle_dataset.py [hunk 6]
```diff
@@ -90,7 +93,7 @@ def __init__(
         project_name,
         bucket_name,
         config=Config(),
-        data_format_fn=lambda data: data,
+        data_format_fn=None,
```
Comment (jdnurme) [inline dataflux_pytorch/dataflux_mapstyle_dataset.py:96]: Why is this necessary?

File: dataflux_pytorch/dataflux_mapstyle_dataset.py [hunk 7]
```diff
@@ -140,23 +150,26 @@ def __getitem__(self, idx):
...
                 storage_client=self.storage_client,
-                dataflux_download_optimization_params=self.
-                dataflux_download_optimization_params,
+                dataflux_download_optimization_params=self.dataflux_download_optimization_params,
                 threads=self.config.threads_per_process,
                 retry_config=self.config.download_retry_config,
             )
         ]
 
+    def data_format_default(self, data):
```
Comment (jdnurme) [inline dataflux_pytorch/dataflux_mapstyle_dataset.py:170]: Even if we do need to have this as a separate function, i don't see any reason for it to be a member of the class. Can we define it outside of the class scope?

File: dataflux_pytorch/dataflux_mapstyle_dataset.py (hunk 6 above)
Comment (akansha1812) [reply dataflux_pytorch/dataflux_mapstyle_dataset.py:96]: Multiprocessing in pytorch dataloader cannot pickle lambda functions. 
Getting error: 
"_pickle.PicklingError: Can't pickle <function DataFluxIterableDataset.<lambda> at 0x10a623550>: attribute lookup DataFluxIterableDataset.<lambda> on dataflux_pytorch.dataflux_iterable_dataset failed"

File: dataflux_pytorch/dataflux_mapstyle_dataset.py (hunk 7 above)
Comment (akansha1812) [reply dataflux_pytorch/dataflux_mapstyle_dataset.py]: Done

File: dataflux_pytorch/tests/test_dataflux_mapstyle_dataset.py [hunk 8]
```diff
@@ -284,8 +284,6 @@ def test_init_sets_user_agent(self):
             self.want_objects,
             f"got listed objects {ds.objects}, want {self.want_objects}",
         )
-        self.assertTrue(
-            self.storage_client._connection.user_agent.startswith("dataflux"))
```
Comment (MattIrv) [inline dataflux_pytorch/tests/test_dataflux_mapstyle_dataset.py:288]: Is the user-agent still getting set somewhere further down the stack?

**Review by MattIrv — APPROVED** [review]

File: dataflux_pytorch/dataflux_mapstyle_dataset.py [hunk 9]
```diff
@@ -115,10 +124,11 @@ def __init__(
             None.
         """
         super().__init__()
+        if storage_client is not None and OS != LINUX:
```
Comment (MattIrv) [inline dataflux_pytorch/dataflux_mapstyle_dataset.py:127]: For these checks, rather than checking the OS, would it make more sense to check the set multiprocessing start_method instead? I think that would be more reliable, since even Linux users might use "spawn" in some cases https://docs.python.org/3/library/multiprocessing.html#multiprocessing.get_start_method

**Review by MattIrv — APPROVED** [review]

**Review by MattIrv — APPROVED** [review]
